### PR TITLE
code optimization : use make build instead of go build on docker multi-stage build

### PR DESCRIPTION
**Pull request by wuyingjun-lucky**
Signed-off-by: wuyingjun <wuyingjun_yewu@cmss.chinamobile.com>

**What type of PR is this?**

**What this PR does / why we need it**:
Use build defined in makefile instead of go build process on docker build stage 
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```NONE

```

### CODE REVIEW

File: cluster/images/karmada-agent/Dockerfile [hunk 1]
```diff
@@ -1,14 +1,10 @@
-FROM golang:1.17-alpine3.15 AS builder
```
Comment (RainbowMango) [inline cluster/images/karmada-agent/Dockerfile:1]: just question:
What's the difference between @golang:1.17-alpine3.15 and golang:1.17 ? Why do you think `golang:1.17` is better?

Comment (wuyingjun-lucky) [reply cluster/images/karmada-agent/Dockerfile:1]: 1: whatever golang:1.17-alpine3.15 or golang:1.17 ,they are just build environment which were not put into the final image
2: while go1.17 can execute make comand but alpine can not

**Comment by RainbowMango** [comment]
cc @gy95 Could you please help with it?

**Comment by gy95** [comment]
seems OK, but when i run make image-karmada-aggregated-apiserver, there're one 
`make: hack/version.sh: Permission denied` in red color maybe we should chmod hack/version.sh to 755

**Comment by wuyingjun-lucky** [comment]
> seems OK, but when i run make image-karmada-aggregated-apiserver, there're one `make: hack/version.sh: Permission denied` in red color maybe we should chmod hack/version.sh to 755

Yes ,I have fixed it  in https://github.com/karmada-io/karmada/pull/1701

**Comment by RainbowMango** [comment]
> Yes ,I have fixed it in https://github.com/karmada-io/karmada/pull/1701

> make: hack/version.sh: Permission denied
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build \
	-ldflags '' \

Seems this PR didn't base on the #1701.  Can you help rebase the PR?

**Comment by RainbowMango** [comment]
You can do that by :
```
$ git checkout code_optmization
$ git fetch upstream
$ git rebase upstream/master
$ git push origin code_optmization -f 
```

**Comment by wuyingjun-lucky** [comment]
> You can do that by :
> 
> ```
> $ git checkout code_optmization
> $ git fetch upstream
> $ git rebase upstream/master
> $ git push origin code_optmization -f 
> ```

done 
But https://github.com/karmada-io/karmada/pull/1701 is only a bug fix in go build stage in make 
what is the reason to rebase

**Comment by RainbowMango** [comment]
> what is the reason to rebase

I found it from [CI logs](https://github.com/karmada-io/karmada/runs/6246361525?check_suite_focus=true#step:4:52). And I'm testing this patch on my side too.

**Comment by wuyingjun-lucky** [comment]
> > what is the reason to rebase
> 
> I found it from [CI logs](https://github.com/karmada-io/karmada/runs/6246361525?check_suite_focus=true#step:4:52). And I'm testing this patch on my side too.

Got it  3q

**Comment by RainbowMango** [comment]
> fatal: No names found, cannot describe anything.
fatal: No names found, cannot describe anything.

I found these errors from [the log](https://github.com/karmada-io/karmada/runs/6303386723?check_suite_focus=true#step:4:47), do you know why?

**Comment by wuyingjun-lucky** [comment]
> I guess we need to check out the code with full `fetch-depth` for [e2e jobs](https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/.github/workflows/ci.yml#L89-L90), just like what we did for [compile job](https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/.github/workflows/ci.yml#L58-L61).

I think u are right 

> I guess we need to check out the code with full `fetch-depth` for [e2e jobs](https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/.github/workflows/ci.yml#L89-L90), just like what we did for [compile job](https://github.com/karmada-io/karmada/blob/dd75a65f1014460652018e60e0cbf4c74a694da4/.github/workflows/ci.yml#L58-L61).

I guess u are right , I will have a test

**Comment by wuyingjun-lucky** [comment]
It works.  please help to review again @RainbowMango

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**